### PR TITLE
feat(api): Publish `OrganizationAlertRuleIndexEndpoint`

### LINKED ISSUE #56676: Publish `OrganizationAlertRuleIndexEndpoint`
We need to publish the GET and POST endpoints at `src/sentry/incidents/endpoints/organization_alert_rule_index.py` to allow users to get a list of metric alert rules for an organization, as well as create them via API request.

❗ Currently on hold until Metric Alert Rule types are documented.

### PR DESCRIPTION
**Pull request by isabellaenriquez**
This PR publishes the `GET` and `POST` endpoints.

Closes #56676 

~~❗ Blocked by #55956~~
~~❗ Blocked by https://github.com/getsentry/sentry-docs/pull/8027~~
~~❗ Blocked by https://github.com/getsentry/sentry-docs/pull/8127~~

# List Metric Alert Rules
localhost_3000_api_alerts_create-a-metric-alert-rule-for-an-organization_ (1)

# Create a Metric Alert Rule
localhost_3000_api_alerts_create-a-metric-alert-rule-for-an-organization_

### CODE REVIEW

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 1]
```diff
@@ -13,18 +15,30 @@
...
+from sentry.apidocs.constants import RESPONSE_FORBIDDEN, RESPONSE_NOT_FOUND, RESPONSE_UNAUTHORIZED
+from sentry.apidocs.examples.metric_alert_examples import MetricAlertExamples
+from sentry.apidocs.parameters import GlobalParams
+from sentry.apidocs.utils import inline_sentry_response_serializer
 from sentry.constants import ObjectStatus
 from sentry.incidents.logic import get_slack_actions_with_async_lookups
 from sentry.incidents.models import AlertRule, Incident
 from sentry.incidents.serializers import AlertRuleSerializer
+
+# from sentry.incidents.serializers.alert_rule import AlertRuleSerializerResponse
```
Comment (sentaur-athena) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:41]: nit: remove

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 2]
```diff
@@ -247,22 +261,327 @@ def get(self, request: Request, organization) -> Response:
         return response
 
 
+class OrganizationAlertRuleIndexPostSerializer(serializers.Serializer):
+    name = serializers.CharField(max_length=64, help_text="The name for the rule.")
```
Comment (sentaur-athena) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:265]: The 64 doesn't show up in docs, should we add it to the help text?

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 3]
```diff
@@ -247,22 +261,327 @@ def get(self, request: Request, organization) -> Response:
...
+    aggregate = serializers.CharField(
+        help_text="A string representing the aggregate used in this alert rule."
```
Comment (sentaur-athena) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:267]: what are valid aggregates?

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 4]
```diff
@@ -247,22 +261,327 @@ def get(self, request: Request, organization) -> Response:
...
+            401: RESPONSE_UNAUTHORIZED,
+            403: RESPONSE_FORBIDDEN,
+            404: RESPONSE_NOT_FOUND,
+        },
+        examples=MetricAlertExamples.LIST_METRIC_ALERT_RULES,  # TODO: make
+    )
     def get(self, request: Request, organization) -> Response:
         """
-        Fetches metric alert rules for an organization
+        Return a list of active metric alert rules bound to an organization.
```
Comment (sentaur-athena) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:390]: nit: Returns? 
I see actually we're saying Return everywhere but shouldn't this be what the endpoint does? which is returns? @shanamatthews 😅

**Review by sentaur-athena — COMMENTED** [review]
Just a few comments. Looks good from apis side otherwise.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py (hunk 3 above)
Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:267]: The documentation for Create a Metric Alert Rule starts with all the different types of metric alert rules we let you create, which includes the aggregates you need to use for each one! I can add a note here to go back to that section for valid aggregates if we think it's necessary?

Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:267]: oh wait i realized i listed the valid values for `datasets` anyway even though they were listed in the different types of metric alert rules. I'll add all the valid aggregates in this section as well!

File: src/sentry/api/serializers/models/alert_rule.py [hunk 5]
```diff
@@ -1,7 +1,8 @@
 from collections import defaultdict
-from typing import MutableMapping
+from typing import List, MutableMapping, Optional
```
Comment (armenzg) [inline src/sentry/api/serializers/models/alert_rule.py:2]: Can you please add `from __future__ import annotations` and then use `list`?

File: src/sentry/api/serializers/models/alert_rule.py
Comment (armenzg) [inline src/sentry/api/serializers/models/alert_rule.py:1]: Please remove `sentry.api.serializers.models.alert_rule` from `pyproject.toml` or the typing changes won't be meaningful.

**Review by armenzg — CHANGES_REQUESTED** [review]
Please separate the typing changes into a different PR to help this PR be focused.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 6]
```diff
@@ -247,22 +259,333 @@ def get(self, request: Request, organization) -> Response:
         return response
 
 
+class OrganizationAlertRuleIndexPostSerializer(serializers.Serializer):
+    name = serializers.CharField(
+        max_length=64,
+        help_text="The name for the rule, which has a maximimum length of 64 characters.",
+    )
+    aggregate = serializers.CharField(
+        help_text="A string representing the aggregate used in this alert rule. Valid aggregate functions are `count`, `count_unique`, `percentage`, `avg`, `apdex`, `failure_rate`, `p50`, `p75`, `p95`, `p99`, `p100`, and `percentile`. See [Metric Alert Rule Types](#metric-alert-rule-types) for valid configurations."
```
[the remark this answers is not in the document]
Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:268]: See https://develop.sentry.dev/api/public/#5-method-decorator

File: src/sentry/api/serializers/models/alert_rule.py (hunk 5 above)
Comment (isabellaenriquez) [reply src/sentry/api/serializers/models/alert_rule.py:2]: To my understanding `List` is used more commonly within `src/sentry/api`

File: src/sentry/api/serializers/models/alert_rule.py [hunk 7]
```diff
@@ -24,6 +25,44 @@
 from sentry.snuba.models import SnubaQueryEventType
 
 
+class AlertRuleSerializerResponseOptional(TypedDict, total=False):
+    environment: Optional[str]
```
[the remark this answers is not in the document]
Comment (isabellaenriquez) [reply src/sentry/api/serializers/models/alert_rule.py:29]: Similar to `List`, I was following conventions being used in similar files. Is there a reason to use `<some_type> | None` over `Optional[type]`?

**Review by armenzg — APPROVED** [review]
You can follow up on removing files from pyproject.toml in another PR.

File: src/sentry/api/serializers/models/alert_rule.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/api/serializers/models/alert_rule.py:30]: note that until we're on python 3.9 we can't use the `from __future__ import annotations` type stylings.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 8]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+        values that determine when an alert should be triggered. Metric alert rules are used to monitor
+        and notify you when certain metrics, such as error count, latency, or failure rate, cross a
+        predefined threshold. These rules help you proactively identify and address issues in your
+        application or system.
+
+        ## Metric Alert Rule Types
+        Listed below are the types of metric alert rules you can create with Sentry, and the values they
+        require. See [Alert Types](/product/alerts/alert-types/#metric-alerts) for more details on each
+        metric alert rule type. All other parameters can be customized depending on how you want the alert
+        rule to work. Scroll down to Body Parameters for more information.
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:434]:
```suggestion
        Below are the types of metric alert rules you can create and the parameter values required
         to set them up. All other parameters can be customized based on how you want the alert
        rule to work. Scroll down to Body Parameters for more information. Visit the 
        [Alert Types](/product /alerts/alert-types/#metric-alerts) docs for more details on each
        metric alert rule type.
```

Comment (shanamatthews) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:434]: Ideally, we should put anchors in for each parameters section so we can link to them, but I'll worry about that for another time!

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 9]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+
+        ### [Number of Errors](/product/alerts/alert-types/#number-of-errors)
+        - `eventTypes`: Any of `error` or `default`.
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:437]: Do we explain what "default" is anywhere? I didn't see it.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 10]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
 
 
+@extend_schema_serializer(exclude_fields=["excludedProjects", "thresholdPeriod"])
+class OrganizationAlertRuleIndexPostSerializer(serializers.Serializer):
+    name = serializers.CharField(
+        max_length=64,
+        help_text="The name for the rule, which has a maximimum length of 64 characters.",
+    )
+    aggregate = serializers.CharField(
+        help_text="A string representing the aggregate used in this alert rule. Valid aggregate functions are `count`, `count_unique`, `percentage`, `avg`, `apdex`, `failure_rate`, `p50`, `p75`, `p95`, `p99`, `p100`, and `percentile`. See [Metric Alert Rule Types](#metric-alert-rule-types) for valid configurations."
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:269]:
```suggestion
        help_text="A string representing the aggregate function used in this alert rule. Valid aggregate functions are `count`, `count_unique`, `percentage`, `avg`, `apdex`, `failure_rate`, `p50`, `p75`, `p95`, `p99`, `p100`, and `percentile`. See [Metric Alert Rule Types](#metric-alert-rule-types) for valid configurations."
```

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 11]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+    {
+        "label": "warning",
+        "alertThreshold": 25,
+        "actions": []
+    }
+]
+```
+Metric alert rule trigger actions follow the following structure:
+- `type`: The type of trigger action. Valid values are `email`, `slack`, `msteams`, `pagerduty`, `sentry_app`, `sentry_notification`, and `opsgenie`.
+- `targetType`: The type of target the notification will be sent to. Valid values are `specific`, `user`, `team`, and `sentry_app`.
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:328]: It's not obvious to me what the "specific" option would do - should we add a short explanation?

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 12]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+- `targetIdentifier`: The ID of the target. This is required as an integer for PagerDuty and Sentry apps, and as a string for all others. Examples of appropriate values include a Slack channel name (`#my-channel`), a user ID, a team ID, a Sentry app ID, etc.
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:329]:
```suggestion
- `targetIdentifier`: The ID of the target. This must be an integer for PagerDuty and Sentry apps, and a string for all others. Examples of appropriate values include a Slack channel name (`#my-channel`), a user ID, a team ID, a Sentry app ID, etc.
```

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 13]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+        help_text="The `SnubaQuery.Type` of the query. If no value is provided, `queryType` is set to the default for the specified `dataset.` See [Metric Alert Rule Types](#metric-alert-rule-types) for valid configurations.",
+    )
+    eventTypes = serializers.ListField(
+        child=serializers.CharField(),
+        required=False,
+        help_text="List of event types that this alert will be related to. Valid values are `default`, `error` and `transaction`.",
+    )
+    comparisonDelta = serializers.IntegerField(
+        required=False,
+        help_text='An optional int representing the time delta to use to determine the comparison period, in minutes. Required when using a percentage change threshold ("x%" higher or lower compared to `comparisonDelta` minutes ago). A percentage change threshold cannot be used for [Crash Free Session Rate](#crash-free-session-rate) or [Crash Free User Rate](#crash-free-user-rate).',
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:356]: Is this change still correct? Trying to simplify the wording.

```suggestion
        help_text='An optional int representing the time delta to use as the comparison period, in minutes. Required when using a percentage change threshold ("x%" higher or lower compared to `comparisonDelta` minutes ago). A percentage change threshold cannot be used for [Crash Free Session Rate](#crash-free-session-rate) or [Crash Free User Rate](#crash-free-user-rate).',
```

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 14]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+    )
+    resolveThreshold = serializers.FloatField(
+        required=False,
+        help_text="Optional value that the subscription needs to reach to resolve the alert. Set automatically if no value is provided. If `thresholdType` is `0`, `resolveThreshold` must be greater than the critical threshold, otherwise, it must be less than the critical threshold.",
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:360]: I wasn't sure what "subscription" meant here, so I changed to "metric". 

Might also be good to give more info on what it is set to automatically if not provided.

```suggestion
        help_text="Optional value that the metric needs to reach to resolve the alert. Set automatically if no value is provided. If `thresholdType` is `0`, `resolveThreshold` must be greater than the critical threshold, otherwise, it must be less than the critical threshold.",
```

**Review by shanamatthews — APPROVED** [review]
Mostly nits, but there are a few spots where I think additional info could help!

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py (hunk 13 above)
Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:356]: that makes sense!

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py (hunk 11 above)
Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:328]: will add!

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py (hunk 9 above)
Comment (isabellaenriquez) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py:437]: Adding a description! don't think we actually explain what `default` event type means anywhere in the docs 😅

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 15]
```diff
@@ -247,22 +259,311 @@ def get(self, request: Request, organization) -> Response:
...
+    )
+    comparisonDelta = serializers.IntegerField(
+        required=False,
+        help_text='An optional int representing the time delta to use as the comparison period, in minutes. Required when using a percentage change threshold ("x%" higher or lower compared to `comparisonDelta` minutes ago). A percentage change threshold cannot be used for [Crash Free Session Rate](#crash-free-session-rate) or [Crash Free User Rate](#crash-free-user-rate).',
+    )
+    resolveThreshold = serializers.FloatField(
+        required=False,
+        help_text="Optional value that the metric needs to reach to resolve the alert. Set automatically based on the lowest severity trigger's `alertThreshold` if no value is provided. For example, if the alert is set to trigger at the warning level when the number of errors is above 50, then the alert would be set to resolve when there are less than 50 errors. If `thresholdType` is `0`, `resolveThreshold` must be greater than the critical threshold, otherwise, it must be less than the critical threshold.",
```
Comment (shanamatthews) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py:360]:
```suggestion
        help_text="Optional value that the metric needs to reach to resolve the alert. If no value is provided, thisSet automatically based on the lowest severity trigger's `alertThreshold` if no value is provided. For example, if the alert is set to trigger at the warning level when the number of errors is above 50, then the alert would be set to resolve when there are less than 50 errors. If `thresholdType` is `0`, `resolveThreshold` must be greater than the critical threshold, otherwise, it must be less than the critical threshold.",
```

**Review by shanamatthews — APPROVED** [review]
